File: gatelogue-types-rs/src/node/town.rs
use strum_macros::EnumString;

use crate::{_from_sql_for_enum, _get_column, node_type, util::ID};

#[derive(Clone, Copy, PartialEq, Eq, Debug, EnumString)]
pub enum Rank {
    Unranked,
    Councillor,
    Mayor,
    Senator,
    Governor,
    Premier,
    Community,
}
_from_sql_for_enum!(Rank);

node_type!(located Town);
impl Town {
    _get_column!("Town", name, String);
    _get_column!("Town", rank, Rank);
    _get_column!("Town", mayor, String);
    _get_column!("Town", deputy_mayor, "deputyMayor", Option<String>);
}

File: gatelogue-types-ts/src/town.ts
import { LocatedNode } from "./located.js";

export type Rank =
  | "Unranked"
  | "Councillor"
  | "Mayor"
  | "Senator"
  | "Governor"
  | "Premier"
  | "Community";

export class Town extends LocatedNode {
  get name(): string {
    return this.getColumn("Town", "name");
  }
  get rank(): Rank {
    return this.getColumn("Town", "rank");
  }
  get mayor(): string {
    return this.getColumn("Town", "mayor");
  }
  get deputyMayor(): string | null {
    return this.getColumn("Town", "deputyMayor");
  }
}
